### DISCUSSION TITLE
EOS Webcam Utility

### DISCUSSION THREAD

**Discussion by raresn** [airsdk/Adobe-Runtime-Support#3062] [Ideas]
Hey everyone!
We have a project where we need to use the EOS Webcam Utility as a webcam, but failing
Any suggestions?

![WhatsApp Image 2024-02-09 at 17 26 06_b81ddc0f](https://github.com/airsdk/Adobe-Runtime-Support/assets/1768694/c3a2deef-aa69-43ba-b768-da38fd095215)

`package {
    
    import flash.display.Sprite;
    import flash.events.PermissionEvent;
    import flash.media.Camera;
    import flash.media.Video;
    import flash.text.TextField;
    import flash.text.TextFieldAutoSize;
    import flash.events.StatusEvent;
    
    [SWF(frameRate="60", backgroundColor="#ffffff", width="1200", height="960")]
    public class CameraApp extends Sprite {
        private var cameraText:TextField;
        private var video:Video;
        private var cameras:Array;
        
        public function CameraApp() {
            init();
        }
        
        private function init():void {
            cameraText = new TextField();
            cameraText.width = 400;
            cameraText.height = 1000;
            cameraText.autoSize = TextFieldAutoSize.LEFT;
            addChild(cameraText);
            
            video = new Video(640, 480);
            video.x = 400;
            addChild(video);
            
            try {
                cameras = Camera.names;
                cameraText.text = "Available Cameras:\n" + cameras.join("\n");
                
                if (cameras.length > 0) {
                    var camera:Camera = Camera.getCamera(String(0));
                    if (camera != null) {
                        cameraText.appendText("\nHave Camera");
                        camera.setQuality(0, 100);
                        camera.addEventListener(StatusEvent.STATUS, onCameraStatus);
                        camera.addEventListener(PermissionEvent.PERMISSION_STATUS, onCameraPermission);
                        camera.setMode(1280, 720, 15);
                        
                        cameraText.appendText("\nn:" + camera.name);
                        cameraText.appendText("\nw:" + camera.width);
                        cameraText.appendText("\nh:" + camera.height);
                        cameraText.appendText("\nf:" + camera.fps);
                        cameraText.appendText("\nc:" + camera.currentFPS);
                        cameraText.appendText("\np:" + camera.position);
                        cameraText.appendText("\nb:" + camera.bandwidth);
                        
                        cameraText.appendText("\n" + Camera.permissionManager.permissionStatus);
                        
                        video.attachCamera(camera);
                    } else {
                        cameraText.appendText("\nNo camera available.");
                    }
                } else {
                    cameraText.appendText("\nNo camera found.");
                }
            } catch (e:Error) {
                cameraText.text = "Error: " + e.message;
            }
        }
        
        private function onCameraPermission(event:PermissionEvent):void
        {
            cameraText.appendText("\nCamera permission: " + event.status);
        }
        
        private function onCameraStatus(event:StatusEvent):void
        {
            cameraText.appendText("\nCamera status: " + event.code);
        }
    }

}`

**Comment by ajwfrost** [comment]
Hi

Forgive my question, but do you have a camera connected to the computer when you try to run this? IIRC you would normally see a placeholder screen from the utility, unless/until it detects a camera... is there a camera already connected when you start up?

I'm wondering what modes it advertises though.. seems like this isn't exposed to ActionScript, but internally the runtime goes through and requests the modes that are supported by a camera provider; maybe in this case there aren't any until a camera is connected.

We can try to reproduce and see what happens...

thanks

**Reply by Adrian-S** [reply]
We did not have a camera connected when testing. But we get the same result on clients computer with a camera connected.
We have tried different software to create a proxy:
`OBS Virtual Camera` - OBS Studio, same issue is present while trying to use this one. `EOS Webcam Utility` is visible in OBS.
`Intel Virtual Camera` is related to Intel Arc video drivers, but I can't configure because it seams it's always in use.
`IP Camera [JPEG/MJPEG]` is WebcamXP... an old software, and is visible in Air. WebcamXP has the same issue trying to connect to EOS or OBS, but works with `Intel Virtual Camera`.

We have also tested with webcamtests dot com, and EOS works.

Using Camera app from windows works as well with EOS.

This was compiled as a bundle, with IntelliJ and auto generated XML.
Could this be related to architecture (x86 / x64) ? Or just old camera libraries?

**Comment by Townado** [comment]
@raresn 

What happens if you try `if (cameras.length > 0) {
                var camera:Camera = Camera.getCamera();` in your code instead?

**Reply by Adrian-S** [reply]
`Camera.getCamera(String(0));` is equivalent to `Camera.getCamera();` and was the 1st version we tried.
The "trace" indicates we have the correct camera. We've also tested with OBS `Camera.getCamera(String(3));` in our case, and has the same issue.

**Comment by raresn** [comment]
@ajwfrost did you manage to reproduce?

**Reply by ajwfrost** [reply]
Hi - sorry, yes, although without a camera attached. I think we'll be able to test it with a camera in a week or so..

But currently: it's hitting an issue when trying to set up the internal modes, which results in us not having any supported modes. Hence the weird return values from the properties for width/height/framerate/etc.

The issue is just that we're unable to connect the camera to our renderer.. IFilterGraph::ConnectDirect is failing to find any way to connect up the camera utility. I would hope that when there's a camera connected, this then starts to work, but it may not I guess, based on your testing.

The exact same thing happens with "OBS Virtual Camera", which makes me think we can look into this further.. but I'm also aware that the IFilterGraph mechanism - and the whole Win32 camera implementation - is based on DirectShow which is now legacy. It might be simplest for us to look instead to update the camera functionality to use Media Foundation classes. First step perhaps being to ensure that it's possible to get the output of these virtual webcams using that..!

thanks

**Reply by ramraver** [reply]
As far as I experienced it, Air fails to access the virtual cam of the 28.x OBS versions. The same applies for Magewell capture cards, though not sure if for all. We had that issue with Knossi using our software while a Magewell capture card was installed. Aditionally all video sources above the indices of the Magewell were blocked also (solved with a workaround at that time by disabling the Magewell in system properties and using another one). Prior to that version, where major things got a complete overhaul, no problems. This relates somehow to the post I made earlier today.

**Reply by raresn** [reply]
@ajwfrost the Media Foundation classes sounds good - is it a huge job?

**Reply by ramraver** [reply]
true. and it would also imply a screencapture/recording feature...

**Reply by ajwfrost** [reply]
Hi

Yes we have this same issue with the OBS as well as Canon/EOS, where the DirectShow filter graph connection is failing. These camera connections work from within e.g. MS Teams, where even though there's no cameras connected, we get a preview still image showing a status screen. So there's presumably something we can do to get these working...

For Media Foundation classes, it's not a massive job I think, as they look like they integrate fairly well and are a similar structure/set-up to the DirectShow media sources. A quick test hack and we have the list of available cameras, it seems relatively straightforward to also get the video/image stream.

The main issue though with Media Foundation is that it doesn't seem that OBS Virtual Webcam supports this! The OBS options does not appearing in the list...  and it does seem as if OBS only supports DirectShow still - found this:
https://medium.com/deelvin-machine-learning/how-does-obs-virtual-camera-plugin-work-on-windows-e92ab8986c4e

With a bit of help from that .. it looks like our initial direct connection fails because OBS supports NV12 and we are looking for RGB, but we then try to connect up a conversion filter. An enumeration based on those types results in a couple of options, but the connection between the OBS virtual camera's output and the conversion filter's input is failing again...

Getting the media types: https://github.com/Fenrirthviti/obs-virtual-cam/blob/136725a82b98d44e3d29088b90e1817625724217/src/virtual-source/virtual-cam.cpp#L418
Checking the connection: https://github.com/Fenrirthviti/obs-virtual-cam/blob/136725a82b98d44e3d29088b90e1817625724217/src/virtual-source/virtual-cam.cpp#L294

I don't think we're seeing the above code in action: we are seeing three configurations (rather than the four being set up there):
1) w 0x780, h 0x438, time per frame 0x28B0A, subtype {3231564E-0000-0010-8000-00AA00389B71} = MEDIASUBTYPE_NV12
2) w 0x780, h 0x438, time per frame 0x28B0A, subtype {30323449-0000-0010-8000-00AA00389B71} = MEDIASUBTYPE_I420
3) w 0x780, h 0x438, time per frame 0x28B0A, subtype {32595559-0000-0010-8000-00AA00389B71} = MEDIASUBTYPE_YUY2
But the interesting thing is that we try to connect via a conversion filter, every one of those is failing to connect to the OBS pin, even though it's meant to have the capabilities to convert from each of those media sub-types...

We could of course accept YUV2 data and convert it ourselves .. but even when we accept this - and the filter graph does then connect and start running - we don't then get any data from the OSB filter, even when it's got video output...

So I think first we need to work on the existing DirectShow mechanisms, to make it more robust and to work with OBS...

For Media Foundation, probably the better option is to use the new implementation just with the new AS3 Media APIs. So we will also look at that in case it helps with the EOS webcam utility, but it sounds like this then won't work for OBS.

thanks

**Comment by raresn** [comment]
@ajwfrost ok, so we will wait for you to do a test with the new AS3 Media API (which i am not familiar with, not sure what this is really) and let us know when/how to do a test ourselves? (We only care about the EOS Webcam Utility at this point)

**Reply by ajwfrost** [reply]
Well .. actually, we have managed to get the EOS Webcam Utility to work now it seems.. if we ignore the fact that we're unable to set up a connection directly, then we fall back into a mode that uses a 'capture builder' component to try to build up the DirectShow pipeline, and with a slight tweak to that, we're then getting the camera preview working...

Still no joy trying to get the OBS virtual camera working though - it also fails when we try it with Microsoft's DirectShow capture sample ("playcap") - and again, doesn't appear when we use their Media Foundation sample.

So for this EOS Webcam utility, we can get a fix out in the short term, we'd just need to ensure the configuration/fallback works properly across different cases. For the longer term we'll provide these capabilities via the Media API but unless/until OBS update their component, it won't be possible to use that one. (FYI the media API isn't actually published yet .. it's in the airglobal.swc definitions but not on the documentation website... still having to do a few refinements to it!)

**Reply by raresn** [reply]
Sounds very good!
Thank you for looking into this so fast. Do you have a time horizon where we could receive the fix? Just trying to plan things on my end.

**Reply by ajwfrost** [reply]
We've submitted this change into the AIR 51 branch - the initial version of AIR 51 failed the QA testing so we're about to cut another one, so when that beta comes out, it would be worth checking that scenario! Depending on timescales we would then either push out 50.2.4.5 with any other fixes in it, or just bring out AIR 51 as a proper release... so within 3-4 weeks I think we should have this in a proper release.

thanks

**Reply by raresn** [reply]
Thanks a lot! Looking forward to test.
